config/v1/types_infrastructure.go

```diff
@@ -123,6 +123,17 @@ const (
 	ExternalTopologyMode TopologyMode = "External"
 )
 
+// PlatformLoadBalancerType defines the type of load balancer used by the cluster.
+type PlatformLoadBalancerType string
+
+const (
+	// LoadBalancerTypeUserManaged is a loadbalancer with control-plane VIPs managed outside of the cluster by the customer.
+	LoadBalancerTypeUserManaged PlatformLoadBalancerType = "UserManaged"
+
+	// LoadBalancerTypeOpenShiftManaged is a loadbalancer with control-plane VIPs managed by the OpenShift cluster.
+	LoadBalancerTypeOpenShiftManaged PlatformLoadBalancerType = "OpenShiftManaged"
```

Comment:
If we revisit the idea of a BGP-managed VIP, which seems quite likely, what would we call it?

Reply:
we would probably have an API to define a BGP configuration (in the FailureDomains?) and this config would be exposed in install-config, that we can check on. WDYT?

Reply:
BTW I'm not against your idea Matt, I'll ask @patrickdillon and @JoelSpeed what they think.

Reply:
This is coming back to what do we call this that isn't `external` or `internal`, realistically, `OpenShiftManaged `and `internal` would both represent both the keepalived and BGP based solutions. 

I did mention somewhere the idea of using something that suggested about the implementation (eg a value of `OpenShiftKeepAlived`, but I'm not sure if exposing the implementation details is desired)

Do you imagine in the future that some operator internal to the openshift cluster would want a value in this enum to say, BGP solution should be enabled now, or are we expecting another configuration to trigger that?

If we want the BGP deploying (or some other system) operator to react to the value of this field, being more specific about the implementation might be useful

Reply:
FailureDomains aren't going to express a loadbalancer configuration. They're about machine configuration.

When we were looking at BGP I think we enumerated them. Lets call them `L2` and `BGP`, which is what MetalLB calls them (kube-vip calls them `ARP` and `BGP`, but it's the same thing). This is a different option, which is no LB at all. I think it's at least not improbable that we'll want to support both these modes in the medium term, and the most obvious place to configure them would be here. They would, however, both be `OpenShiftManaged`.

A problem here is that while `UserManaged` seems like a universal option which could be implemented by all platforms, the nature of and options for `OpenShiftManaged` is likely to vary by platform. Throwing another idea in the mix, what if we wanted to add Octavia support for OpenStack? That's 3.

Just spitballing: how about if this was a boolean? This would select between `UserManaged` and `OpenShiftManaged`, and then we could have a different field to select the nature of `OpenShiftManaged`?

Reply:
@mdbooth would it be fine to rename it to `OpenShiftManagedDefault` ? so later we can add `OpenShiftManagedBGP` and so on. The current stack will remain the default forever I guess... so why not naming it Default?

Reply:
> Just spitballing: how about if this was a boolean? This would select between UserManaged and OpenShiftManaged, and then we could have a different field to select the nature of OpenShiftManaged?

I remember a previous conversation about paired booleans where one only makes sense if the other is set to a certain value and how this caused a lot of confusion around OpenStack APIs 😉 

In theory, you could do what you're suggesting if you're willing to make it a discriminated union, though I think that would only really be needed if you want to add additional configuration options for the different load balancers, otherwise different values for an Enum here would be sufficient right?

```
loadBalancer:
  type: UserManaged | OpenShiftManaged
  openShiftManaged:
    type: Default | BGP | L2 | Octavia
   ... Other fields go here, any that are needed?
```

Reply:
To avoid the scenario where we're blocked in the future by a too restrictive API, I'm going to try what @JoelSpeed proposed and see how that works for everyone.

Reply:
Just to follow up on my previous comment, all I was looking for was consistency between what we call it in the config and what we call it in the code, which it appears the current proposal does.

I do like the option of having implementation-specific config values. BGP would for sure need them and we might even put some things in for L2 if we had that available.

Reply:
I did revisit it and I'm struggling to make this work with proper defaults, I request a bit of help.
I'm basically having openapi errors and I'm sure I messed up something with the defaults but I can't figure out what.
If @JoelSpeed / @deads2k could have a look please.

Thanks

Reply:
I addressed @JoelSpeed 's proposal. I think this is now flexible enough for future use-cases where we want to add more LBs and options to them.